### PR TITLE
Move packet writing code, add graceful shutdown handler, fix close handshake

### PR DESCRIPTION
**Pull request by adam-fowler**
- Move write packet code into `WebSocketOutboundWriter`
- Add cancellation and graceful shutdown handlers
- Implement close process correctly ie send close, wait for responding close

### CODE REVIEW

File: Snippets/WebsocketTest.swift [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
 let router = Router(context: BasicWebSocketRequestContext.self)
 router.middlewares.add(FileMiddleware("Snippets/public"))
 router.get { _, _ in
     "Hello"
 }
 
-router.ws("/ws") { inbound, outbound, _ in
-    for try await packet in inbound {
+router.ws("/ws") { ws, _ in
+    for try await packet in ws.inbound {
```
Comment (Joannis) [inline Snippets/WebsocketTest.swift:15]: I like that this follows the pattern found in NIO. But I wonder if `ws.inbound` as a symbol is easy to reason about for newcomers to the ecosystem.

File: Snippets/WebsocketTest.swift [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
         if case .text("disconnect") = packet {
             break
         }
-        try await outbound.write(.custom(packet.webSocketFrame))
+        try await ws.outbound.write(.custom(packet.webSocketFrame))
```
Comment (Joannis) [inline Snippets/WebsocketTest.swift:19]: `ws.write` would also be easier to reason about IMO

File: Sources/HummingbirdWebSocket/WebSocketHandler.swift [hunk 3]
```diff
@@ -16,96 +16,120 @@ import Hummingbird
...
                         }
+                        try await group.next()
+                        webSocket.inbound.finish()
                     }
-                    group.addTask {
-                        // handle websocket data and text
-                        try await handler(webSocketHandlerInbound, webSocketHandlerOutbound, context)
-                        try await self.close(code: .normalClosure, outbound: outbound, context: context)
+                } onGracefulShutdown: {
+                    Task {
```
Comment (Joannis) [inline Sources/HummingbirdWebSocket/WebSocketHandler.swift:115]: Is there a way around an unstructured task here?

File: Sources/HummingbirdWebSocket/WebSocketHandler.swift [hunk 4]
```diff
@@ -130,62 +155,48 @@ actor WebSocketHandler: Sendable {
...
-        frame.maskKey = self.makeMaskKey()
-        try await outbound.write(frame)
-    }
-
-    /// Make mask key to be used in WebSocket frame
-    private func makeMaskKey() -> WebSocketMaskingKey? {
-        guard self.type == .client else { return nil }
-        let bytes: [UInt8] = (0...3).map { _ in UInt8.random(in: .min ... .max) }
-        return WebSocketMaskingKey(bytes)
+        try await outbound.write(frame: .init(fin: true, opcode: .connectionClose, data: buffer))
```
Comment (Joannis) [inline Sources/HummingbirdWebSocket/WebSocketHandler.swift:185]: I think we should consider making `close` non-throwing. I get that writing a frame can fail, but that's an _acceptable_ and ignorable failure in my book - at this point in the WebSocket's lifecycle. Possibly something to ask the NIO team's opinion on.

File: Sources/HummingbirdWebSocket/WebSocketHandler.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/HummingbirdWebSocket/WebSocketHandler.swift:115]: Unfortunately not. I can't close either the inbound stream or outbound writer as they are needed for the close handshake. It's a recurring issue with graceful shutdown handlers, where something more complex is need to initiate shutdown.

File: Sources/HummingbirdWebSocket/WebSocketHandler.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/HummingbirdWebSocket/WebSocketHandler.swift:185]: I disagree here, this initiates the close handshake and doesn't actually close the websocket. It only gets closed when we receive the close connection message from the client. If we ignore the error then the close connection is never sent and we don't receive a close connection from the client. Remember also this is an internal function. Users will close the connection by exiting the handler.

In the current situation when the error is thrown out of the `asyncChannel.executeThenClose` closure which will force close the connection.

**Review by Joannis — APPROVED** [review]

File: Sources/HummingbirdWebSocket/WebSocketHandler.swift (hunk 4 above)
Comment (Joannis) [reply Sources/HummingbirdWebSocket/WebSocketHandler.swift:196]: Ah, good clarification. I misremember that then

**Pull request merged**